rclcpp/test/rclcpp/test_client.cpp

```diff
@@ -123,3 +171,151 @@ TEST_F(TestClientSub, construction_and_destruction) {
     }, rclcpp::exceptions::InvalidServiceNameError);
   }
 }
+
+class TestClientWithServer : public ::testing::Test
+{
+protected:
+  static void SetUpTestCase()
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  static void TearDownTestCase()
+  {
+    rclcpp::shutdown();
+  }
+
+  void SetUp()
+  {
+    node = std::make_shared<rclcpp::Node>("node", "ns");
+
+    auto callback = [](
+      const test_msgs::srv::Empty::Request::SharedPtr,
+      test_msgs::srv::Empty::Response::SharedPtr) {};
+
+    service = node->create_service<test_msgs::srv::Empty>(service_name, std::move(callback));
+  }
+
+  ::testing::AssertionResult SendEmptyRequestAndWait(
+    std::chrono::milliseconds timeout = std::chrono::milliseconds(1000))
+  {
+    using SharedFuture = rclcpp::Client<test_msgs::srv::Empty>::SharedFuture;
+
+    auto client = node->create_client<test_msgs::srv::Empty>(service_name);
+    if (!client->wait_for_service()) {
+      return ::testing::AssertionFailure() << "Waiting for service failed";
+    }
+
+    auto request = std::make_shared<test_msgs::srv::Empty::Request>();
+    bool received_response = false;
+    auto callback = [&received_response](SharedFuture future_response) {
+        ASSERT_NE(nullptr, future_response.get());
```

Comment:
Should this return `return ::testing::AssertionFailure()` or is it handled it normally with the `ASSERT_NE` macro?

Reply:
Since this line is inside a lambda, this ASSERT_NE has the effect of returning early from the lambda, which means received response won't get set to true. That also means the loop in line 219 will spin for the full second, which is unnecessary.

I fixed this up by setting received_response to true in either case, and just passing through an AssertionResult.